notifier/notifier.go

```diff
@@ -221,43 +220,118 @@ func (n *Notifier) collectRogueCallbacks() {
 	}
 }
 
-// Notify posts callback info to j.CallbackURL
-func (n *Notifier) Notify(j *job.Job) error {
+// MonitorDeliveries starts as many go routines as limit instructs in order to
+// poll/monitor the delivery reports channel.
+func (n *Notifier) MonitorDeliveries(ctx context.Context, limit int, dwg *sync.WaitGroup) {
+	dwg.Add(limit)
+
+	for i := 0; i < limit; i++ {
+		go n.monitorDeliveriesRoutine(ctx, dwg)
+	}
+}
+
+// monitorDeliveriesRoutine watches events from the delivery reports channel and
+// also checks whether the context has been cancelled or not.
+func (n *Notifier) monitorDeliveriesRoutine(ctx context.Context, dwg *sync.WaitGroup) {
+	defer dwg.Done()
+
+	for {
+		select {
+		case <-ctx.Done():
+			n.Log.Println("Context has been cancelled")
+			return
+		case httpInfo := <-n.backends["http"].DeliveryReports():
+			n.handleCallbackInfo(httpInfo)
+			continue
+		case kafkaInfo := <-n.backends["kafka"].DeliveryReports():
+			n.handleCallbackInfo(kafkaInfo)
+			continue
+		}
+	}
+}
+
+// handleCallbackInfo handles a callback info object.
+// If the callback is successful we increment stats counters and remove the job from storage.
+// For kafka if the callback is not successful we immediately mark the callback as failed since
+// kafka has already tried to retransmit a message for an amount of time. For http if the
+// callback is not successful we have already handled this case in Notify() method with retryOrFail().
+// Also not that when on http we call QueueJobForDeletion which schedules file deletion associated to our job,
+// because we are sure the other end has downloaded the file. But kakfa only ensures that a message has been
+// persisted, but most likely consumer has not recieved the message yet alone the file.
+func (n *Notifier) handleCallbackInfo(cbInfo job.CallbackInfo) {
+	j, err := n.Storage.GetJob(cbInfo.JobID)
+	if err != nil {
+		n.Log.Printf("Error: Could not get job %s. Operation returned error: %s", cbInfo.JobID, err)
+	}
+
+	if cbInfo.Success {
+		cbType, _ := j.GetCallbackTypeAndDestination()
+
+		if cbType == "http" {
+			err := n.Storage.QueueJobForDeletion(cbInfo.JobID)
+			if err != nil {
+				n.Log.Printf("Error: Could not queue job %s for deletion: %s", cbInfo.JobID, err)
+			}
+		}
+
+		n.stats.Add(statsSuccessfulCallbacks, 1)
+
+		err := n.Storage.RemoveJob(cbInfo.JobID)
+		if err != nil {
+			n.Log.Printf("Could not remove job %s. Operation returned error: %s", cbInfo.JobID, err)
+		}
+		return
+	}
+
+	n.markCbFailed(&j, cbInfo.Error)
+}
+
+// Prepare runs the necessary actions before calling Notify
+// and returns the payload to be sent.
+func (n *Notifier) Prepare(j *job.Job) ([]byte, error) {
```

Comment:
Name of this function is very generic, conveying no information.

Most important, this function could be reduced to `foo(j *job.Job) (CallbackInfo, err)` and then the `markCbFailed` should be done in the caller, if an error is returned.